File: HOL/dotnet/modern-cloud-apps/end/DevCamp.WebApp/Utils/StorageHelper.cs
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Blob;
using Microsoft.WindowsAzure.Storage.Queue;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Configuration;
using System.IO;
using System.Threading.Tasks;
using System.Web;

namespace DevCamp.WebApp.Utils
{
    public class StorageHelper
    {
        static string account = ConfigurationManager.AppSettings["AZURE_STORAGE_ACCOUNT"];
        static string key = ConfigurationManager.AppSettings["AZURE_STORAGE_ACCESS_KEY"];
        static string AZURE_STORAGE_BLOB_CONTAINER = ConfigurationManager.AppSettings["AZURE_STORAGE_BLOB_CONTAINER"];
        static string AZURE_STORAGE_QUEUE = ConfigurationManager.AppSettings["AZURE_STORAGE_QUEUE"];
        static string blobStorageConnectionString = String.Format("DefaultEndpointsProtocol=https;AccountName={0};AccountKey={1}", account, key);

        /// <summary>
        /// Adds an incident message to the queue
        /// </summary>
        /// <param name="IncidentId">The incident ID from the service</param>
        /// <param name="ImageFileName">The file name of the image</param>
        /// <returns></returns>
        public static async Task AddMessageToQueue(string IncidentId, string ImageFileName)
        {
            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(blobStorageConnectionString);
            CloudQueueClient queueClient = storageAccount.CreateCloudQueueClient();
            CloudQueue msgQ = queueClient.GetQueueReference(AZURE_STORAGE_QUEUE);
            msgQ.CreateIfNotExists();

            JObject qMsgJson = new JObject();
            qMsgJson.Add("IncidentId", IncidentId);
            qMsgJson.Add("BlobContainerName", AZURE_STORAGE_BLOB_CONTAINER);
            qMsgJson.Add("BlobName", getIncidentBlobFilename(IncidentId, ImageFileName));

            var qMsgPayload = JsonConvert.SerializeObject(qMsgJson);
            CloudQueueMessage qMsg = new CloudQueueMessage(qMsgPayload);

            await msgQ.AddMessageAsync(qMsg);
        }

        /// <summary>
        /// Uploads a blob to the configured storage account
        /// </summary>
        /// <param name="IncidentId">The IncidentId the image is associated with</param>
        /// <param name="imageFile">The File</param>
        /// <returns>The Url to the blob</returns>
        public static async Task<string> UploadFileToBlobStorage(string IncidentId, HttpPostedFileBase imageFile)
        {
            string imgUri = string.Empty;

            try
            {
                CloudStorageAccount storageAccount = CloudStorageAccount.Parse(blobStorageConnectionString);

                CloudBlobClient blobClient = storageAccount.CreateCloudBlobClient();
                CloudBlobContainer container = blobClient.GetContainerReference(AZURE_STORAGE_BLOB_CONTAINER);
                container.CreateIfNotExists();
                container.SetPermissions(new BlobContainerPermissions { PublicAccess = BlobContainerPublicAccessType.Blob });

                CloudBlockBlob imgBlob = container.GetBlockBlobReference(getIncidentBlobFilename(IncidentId, imageFile.FileName));
                imgBlob.Properties.ContentType = imageFile.ContentType;
                await imgBlob.UploadFromStreamAsync(imageFile.InputStream);

                var uriBuilder = new UriBuilder(imgBlob.Uri);
                uriBuilder.Scheme = "https";
                imgUri = uriBuilder.ToString();
            }
            catch (Exception ex)
            {
                throw new HttpUnhandledException($"Unable to upload image for incident {IncidentId} to blob storage. Error:: ${ex.ToString()}");
            }
            return imgUri;
        }

        private static string getIncidentBlobFilename(string IncidentId, string FileName)
        {
            string fileExt = Path.GetExtension(FileName);
            //Remove the starting . if exists
            if (fileExt.StartsWith("."))
            {
                fileExt.TrimStart(new char[] { '.' });
            }
            return $"{IncidentId}{fileExt}";
        }
    }
}

File: HOL/java/modern-cloud-apps/end/src/main/java/devCamp/WebApp/Utils/StorageHelper.java
package devCamp.WebApp.Utils;

import org.apache.commons.io.FilenameUtils;
import org.codehaus.jettison.json.JSONException;
import org.codehaus.jettison.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.web.multipart.MultipartFile;
import com.microsoft.azure.storage.*;
import com.microsoft.azure.storage.blob.*;
import com.microsoft.azure.storage.queue.CloudQueue;
import com.microsoft.azure.storage.queue.CloudQueueClient;
import com.microsoft.azure.storage.queue.CloudQueueMessage;

import java.io.*;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.InvalidKeyException;

import javax.ws.rs.core.UriBuilder;

public class StorageHelper {

	//configuration values from the system Environment
	private static String account;
	private static String key;
	private static String AZURE_STORAGE_BLOB_CONTAINER;
	private static String AZURE_STORAGE_QUEUE;
	private static String blobStorageConnectionString;
	
	public static void init() {
		
		account = System.getenv("AZURE_STORAGE_ACCOUNT");;
		key = System.getenv("AZURE_STORAGE_ACCESS_KEY");;
		AZURE_STORAGE_QUEUE = System.getenv("AZURE_STORAGE_QUEUE");
		AZURE_STORAGE_BLOB_CONTAINER = System.getenv("AZURE_STORAGE_BLOB_CONTAINER");
		blobStorageConnectionString = String.format("DefaultEndpointsProtocol=http;AccountName=%s;AccountKey=%s", account,key);
	}

	
	
    /// <summary>
    /// Adds an incident message to the queue
    /// </summary>
    /// <param name="IncidentId">The incident ID from the service</param>
    /// <param name="ImageFileName">The file name of the image</param>
    /// <returns></returns>
    public static void AddMessageToQueue(String IncidentId, String ImageFileName)
    {
    	init();
        CloudStorageAccount storageAccount;
		try {
			storageAccount = CloudStorageAccount.parse(blobStorageConnectionString);
        CloudQueueClient queueClient = storageAccount.createCloudQueueClient();

        CloudQueue msgQ = queueClient.getQueueReference(AZURE_STORAGE_QUEUE);
        msgQ.createIfNotExists();

        JSONObject json = new JSONObject();
        json.append("IncidentId", IncidentId);
        json.append("BlobContainerName", AZURE_STORAGE_BLOB_CONTAINER);
        json.append("BlobName", getIncidentBlobFilename(IncidentId, ImageFileName));

        String msgPayload = json.toString();
        CloudQueueMessage qMsg = new CloudQueueMessage(msgPayload);
        msgQ.addMessage(qMsg);
		} catch (InvalidKeyException | URISyntaxException | StorageException | JSONException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }
	
	public static String UploadFileToBlobStorage(String IncidentId, MultipartFile imageFile){
		init();
		CloudStorageAccount account;
		try {
			account = CloudStorageAccount.parse(blobStorageConnectionString);
			CloudBlobClient serviceClient = account.createCloudBlobClient();

			// Container name must be lower case.
			CloudBlobContainer container = serviceClient.getContainerReference(AZURE_STORAGE_BLOB_CONTAINER);
			container.createIfNotExists();

			// Set anonymous access on the container.
			BlobContainerPermissions containerPermissions = new BlobContainerPermissions();
			containerPermissions.setPublicAccess(BlobContainerPublicAccessType.CONTAINER);
			container.uploadPermissions(containerPermissions);

			// 
			CloudBlockBlob imgBlob = container.getBlockBlobReference(getIncidentBlobFilename(IncidentId,imageFile.getOriginalFilename()));
			imgBlob.getProperties().setContentType(imageFile.getContentType());
//			imgBlob.uploadProperties();
			imgBlob.upload(imageFile.getInputStream(),imageFile.getSize());
			UriBuilder builder = UriBuilder.fromUri(imgBlob.getUri());
			builder.scheme("https");
			return builder.toString();
		} catch (InvalidKeyException | URISyntaxException | StorageException | IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} 
		return null;


	}

	public static String getIncidentBlobFilename(String IncidentId,String FileName) {
		String fileExt = FilenameUtils.getExtension(FileName);
		// TODO check this against the .NET code - 
		//	with this code in, we're generating filenames that don't have a period between the incident id and the extension 
//		if (fileExt.startsWith(".")){
//			fileExt = fileExt.substring(1);
//		}
		return String.format("%s%s", IncidentId,fileExt);
	}
}
